### DISCUSSION TITLE
Changes on @ObservableObject inside State does not update UI

### DISCUSSION THREAD

**Discussion by raduncicen** [pointfreeco/swift-composable-architecture#1630] [Q&A] [answered]
### Description

Hello,
I'm building an ARCamera app. It requires me to wrap ARView to a swiftUI wrapper and provide a coordinator class for it which requires me to hold on to a coordinator class inside the ReducerProtocol's State. However the changes I impose on this ObservableObject do not trigger any update on the UI no matter what i do.

Do you guys have a solution for such cases where we need to use a manager observable class inside the state? 
In other words, is there a way to trigger UI updates for the changes in ObservableObject(aka ARViewCoordinator) that is kept inside the state?

You can see the code below. 

**The ObservableObject that i need to use.**
```
class ARViewCoordinator: NSObject, ObservableObject {
    public init(mode: Camera.CameraMode, layerGroup: LayerGroup) {
        self.mode = mode
        self.layerGroup = layerGroup
    }
    
    var parent: ARViewContainer?
    @Published var mode: Camera.CameraMode // Enum
    @Published var layerGroup: LayerGroup // This is a struct 
    
    func setParent(as parent: ARViewContainer) {
        self.parent = parent
    }
    
    func addLayer(_ layer: ClassicLayer) {
        layerGroup.layers.insert(layer, at: 0)
    }
}

extension ARViewCoordinator: ARSessionDelegate {
    func session(_ session: ARSession, didAdd anchors: [ARAnchor]) {
        //Some logic in here for adding anchors to ARView.
    }
}

```

**The reducer which i use it**
```
public struct Camera: ReducerProtocol {
    public init() {}
    
    public struct State: Equatable {
        @BindableState var arViewCoordinator: ARViewCoordinator
        var testBool: Bool = false // I'm toggling this so that the updates in arViewCoordinator appear in the screen
    }
    
    public enum Action: BindableAction {
        case binding(BindingAction<State>)
        case addNewLayer
    }
    
    public var body: some ReducerProtocol<State, Action> {
        BindingReducer()
        
        Reduce { state, action in
            switch action {
            case .binding(_):
                return .none
            case .addNewLayer:
                let newLayer = ClassicLayer(cameraMode: .classic,
                                            sublayers: .init(subLayers: ["Zoom", "Zoom", "Zoom"], selectedIndex: 0),
                                            opacity: 1,
                                            isHidden: false,
                                            transformation: .identity)
                state.arViewCoordinator.addLayer(newLayer)
//                state.testBool.toggle()
                return .none
            }
        }
    }
}
```

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Destination operating system

IOS 16

### Xcode version information

14.0.1

**Comment by tgrapperon** [comment, marked as answer]
Hey @raduncicen!
If you're using `UIViewRepresentable`, its `Coordinator` is not required to be an `ObservableObject`, and it will furthermore not be observed as such by the `Representable`. When using `UIViewRepresentable`, the `Coordinator` is private, and usually an implementation detail with which you don't talk directly (for example from a reducer).
I never tried to install the `Coordinator` as an `ObservedObject` directly in the `UIViewRepresentable`, so I don't know if it would work.
So the idea is to wrap UIKit in `UIViewRepresentable`, ideally with a value-type API and closures/bindings to control it. But in your case, it could maybe work if you simply rewrite your `View`'s init like this:
```swift
let store: StoreOf<Camera>
@ObservedObject var viewStore: ViewStoreOf<Camera>
@ObservedObject var coordinator: ARViewCoordinator
init(store: StoreOf<Camera>) {
  self.store = store
  let viewStore = ViewStore(store, observe: { $0 })
  self.viewStore = viewStore
  self.coordinator = viewStore.arViewCoordinator
}
```
This is probably not how I would do it, but it can be enough to make it work in you case.

This is likely not a TCA bug, as there is nothing that changes in your case to warrant a new view rendering, so I would say that it's working as expected in this situation.

**Reply by saroar** [reply]
I am also thinking about how to use same login when webView loading to show is page didFinishLoading

**Comment by mbrandonw** [comment]
Hi @raduncicen, as @tgrapperon points out this is not a bug in TCA and so I'm going to convert this to a discussion.

In general, if you hold reference types in `State` (such as you are doing with  `ARViewCoordinator`), then changes to those references types do not cause SwiftUI to update its view. This is true of vanilla SwiftUI too.
